[visionOS] Videos on youtube.com are cropped in compatibility mode when using a mobile UA

pxlcoder opened the pull request:
#### d409ef2d247f82970d9142a9cc6a492aea06d53a
<pre>
[visionOS] Videos on youtube.com are cropped in compatibility mode when using a mobile UA
<a href="https://bugs.webkit.org/show_bug.cgi?id=260192">https://bugs.webkit.org/show_bug.cgi?id=260192</a>
rdar://112444285

Reviewed by Mike Wyrzykowski and Richard Robinson.

Unlike the desktop site, YouTube&apos;s mobile site fullscreens the &lt;video&gt; element
rather than an element containing the &lt;video&gt;. Currently, on visionOS, WebKit
always uses element fullscreen (rather than the native player) when playing
fullscreen &lt;video&gt;. However, this results in cropping on YouTube&apos;s mobile site
as YouTube sets `object-fit: cover` on the &lt;video&gt;. On iPadOS, this CSS would
have no observable effect, as the native player is used.

This issue does not reproduce with YouTube&apos;s mobile site in Safari on visionOS,
even though element fullscreen is enforced, as the fullscreen window&apos;s aspect
ratio matches the &lt;video&gt;.

However, in compatibility mode, the fullscreen behavior should match iPadOS.
Fix by ensuring element fullscreen is not forced for &lt;video&gt; fullscreen in
compatibility mode.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::enterFullscreen):

Use the existing `videoFullscreenRequiresElementFullscreen` to determine whether
to force element fullscreen for &lt;video&gt;. This flag is only true when running on
visionOS, and compatibility mode is not active.

Canonical link: <a href="https://commits.webkit.org/266920@main">https://commits.webkit.org/266920@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a78de19cc780ec275db9670a30aadd9f9211bd4a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15082 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15387 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15739 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16837 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14178 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15220 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17903 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15486 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/16811 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15263 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15725 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12817 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17565 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13005 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13603 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20569 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14083 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13771 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17019 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14333 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/12134 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13614 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17951 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1829 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14174 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -6768,7 +6768,11 @@ void HTMLMediaElement::enterFullscreen(VideoFullscreenMode mode)
 #if ENABLE(FULLSCREEN_API) && ENABLE(VIDEO_USES_ELEMENT_FULLSCREEN)
-    if (document().settings().fullScreenEnabled() && mode == VideoFullscreenModeStandard) {
+    bool videoUsesElementFullscreen = true;
+#if PLATFORM(IOS_FAMILY)
```
I'm guessing this is `PLATFORM(IOS_FAMILY)` in case we set `ENABLE_VIDEO_USES_ELEMENT_FULLSCREEN` to `1` for iOS (as it is currently 0)?

pxlcoder replied:
I used `PLATFORM(IOS_FAMILY)` simply because `videoFullscreenRequiresElementFullscreen` is behind that.

nt1m commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -6768,7 +6768,11 @@ void HTMLMediaElement::enterFullscreen(VideoFullscreenMode mode)
+    bool videoUsesElementFullscreen = true;
+#if PLATFORM(IOS_FAMILY)
+    videoUsesElementFullscreen &= document().settings().videoFullscreenRequiresElementFullscreen();
+#endif
```
```suggestion
#if PLATFORM(IOS_FAMILY)
    bool videoUsesElementFullscreen = document().settings().videoFullscreenRequiresElementFullscreen();
#else
    bool videoUsesElementFullscreen = true;
#endif
```

 I have a slight preference for writing like this.

pxlcoder replied:
Sure.

mwyrzykowski replied:
Oh I actually had the same feedback 👍

mwyrzykowski commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -6768,7 +6768,11 @@ void HTMLMediaElement::enterFullscreen(VideoFullscreenMode mode)
-    if (document().settings().fullScreenEnabled() && mode == VideoFullscreenModeStandard) {
+    bool videoUsesElementFullscreen = true;
+#if PLATFORM(IOS_FAMILY)
+    videoUsesElementFullscreen &= document().settings().videoFullscreenRequiresElementFullscreen();
```
is the point of using `&=` to avoid a compiler warning that the value set to above is unused? Would it be more clear to say this?
```
#if PLATFORM(IOS_FAMILY)
bool videoUsesElementFullscreen = ...;
#else
constexpr bool videoUsesElementFullscreen = true;
#endif
```
but does not matter to me so much

pxlcoder committed:
[visionOS] Videos on youtube.com are cropped in compatibility mode when using a mobile UA
https://bugs.webkit.org/show_bug.cgi?id=260192
rdar://112444285

Reviewed by Mike Wyrzykowski and Richard Robinson.

Unlike the desktop site, YouTube's mobile site fullscreens the <video> element
rather than an element containing the <video>. Currently, on visionOS, WebKit
always uses element fullscreen (rather than the native player) when playing
fullscreen <video>. However, this results in cropping on YouTube's mobile site
as YouTube sets `object-fit: cover` on the <video>. On iPadOS, this CSS would
have no observable effect, as the native player is used.

This issue does not reproduce with YouTube's mobile site in Safari on visionOS,
even though element fullscreen is enforced, as the fullscreen window's aspect
ratio matches the <video>.

However, in compatibility mode, the fullscreen behavior should match iPadOS.
Fix by ensuring element fullscreen is not forced for <video> fullscreen in
compatibility mode.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::enterFullscreen):

Use the existing `videoFullscreenRequiresElementFullscreen` to determine whether
to force element fullscreen for <video>. This flag is only true when running on
visionOS, and compatibility mode is not active.

Canonical link: https://commits.webkit.org/266920@main
